### PR TITLE
[Core][Product] Fix Inexistent method

### PR DESCRIPTION
**Pull request by h-sn**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Related tickets |  |
| License | MIT |

### CODE REVIEW

File: src/Sylius/Component/Core/Model/Product.php [hunk 1]
```diff
@@ -139,7 +139,7 @@ public function getTaxons($taxonomy = null)
     {
         if (null !== $taxonomy) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/Product.php]: `$taxonomy` should be renamed to `$rootTaxonCode`

File: src/Sylius/Component/Core/Model/Product.php [hunk 2]
```diff
@@ -139,7 +139,7 @@ public function getTaxons($taxonomy = null)
...
             return $this->taxons->filter(function (BaseTaxonInterface $taxon) use ($taxonomy) {
-                return $taxonomy === strtolower($taxon->getTaxonomy()->getName());
+                return $taxonomy === strtolower($taxon->getRoot()->getName());
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/Product.php]: `return $rootTaxonCode === $taxon->getRoot()->getCode();`

Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/Product.php]: I think we should also exclude root taxons:
```php
if ($taxon->isRoot()) {
    return false;
}

return $rootTaxonCode === $taxon->getRoot()->getCode();
```

**Comment by michalmarcinkowski** [comment]
Hi @h-sn, thanks for contributing! Are you familiar with PhpSpec? Could you write a spec for this method? (useful links: [PhpSpec docs](http://www.phpspec.net/en/latest/manual/introduction.html), [Sylius contributoring guide](http://docs.sylius.org/en/latest/contributing/code/bdd.html#phpspec))

**Comment by h-sn** [comment]
Need some time

File: src/Sylius/Component/Core/spec/Model/ProductSpec.php [hunk 3]
```diff
@@ -67,6 +68,25 @@ function its_taxons_are_mutable(Collection $taxons)
         $this->getTaxons()->shouldReturn($taxons);
     }
 
+    function its_taxons_should_be_returned_by_taxonomy(
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: `taxonomy` -> `root_taxon_code`

File: src/Sylius/Component/Core/spec/Model/ProductSpec.php [hunk 4]
```diff
@@ -67,6 +68,25 @@ function its_taxons_are_mutable(Collection $taxons)
...
+        TaxonInterface $taxon,
+        TaxonInterface $rootTaxon
+    )
+    {
+        $rootTaxon->addChild($taxon);
+        $rootTaxon->setName('Brand');
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: `$rootTaxon->getName()->willReturn('Brand')`

File: src/Sylius/Component/Core/spec/Model/ProductSpec.php [hunk 5]
```diff
@@ -67,6 +68,25 @@ function its_taxons_are_mutable(Collection $taxons)
         $this->getTaxons()->shouldReturn($taxons);
     }
 
+    function its_taxons_should_be_returned_by_taxonomy(
+        TaxonInterface $taxon,
+        TaxonInterface $rootTaxon
+    )
+    {
+        $rootTaxon->addChild($taxon);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: No need for that.

File: src/Sylius/Component/Core/spec/Model/ProductSpec.php [hunk 6]
```diff
@@ -67,6 +68,25 @@ function its_taxons_are_mutable(Collection $taxons)
...
+        $rootTaxon->setName('Brand');
+
+        $taxon->getRoot()->willReturn($rootTaxon);
+        $taxons = new ArrayCollection([$taxon->getWrappedObject()]);
+
+        $this->setTaxons($taxons);
+        $taxon->getRoot()->shouldBeCalled();
+        $rootTaxon->getName()->shouldBeCalled();
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: After [this](https://github.com/Sylius/Sylius/pull/5170/files#r65704954) we won't need `shouldBeCalled`.

File: src/Sylius/Component/Core/spec/Model/ProductSpec.php [hunk 7]
```diff
@@ -67,6 +68,25 @@ function its_taxons_are_mutable(Collection $taxons)
...
+    )
+    {
+        $rootTaxon->addChild($taxon);
+        $rootTaxon->setName('Brand');
+
+        $taxon->getRoot()->willReturn($rootTaxon);
+        $taxons = new ArrayCollection([$taxon->getWrappedObject()]);
+
+        $this->setTaxons($taxons);
+        $taxon->getRoot()->shouldBeCalled();
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: Redundant

File: src/Sylius/Component/Core/spec/Model/ProductSpec.php [hunk 8]
```diff
@@ -67,6 +68,25 @@ function its_taxons_are_mutable(Collection $taxons)
...
+        $rootTaxon->getName()->shouldBeCalled();
+
+        $this->getTaxons('Brand')
+            ->shouldHaveType('Doctrine\Common\Collections\Collection');
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: `->shouldReturn(new ArrayCollection([$taxon]));`

Comment (h-sn) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: 71  - its taxons should be returned by taxonomy
      expected [obj:Doctrine\Common\Collections\ArrayCollection], but got
      [obj:Doctrine\Common\Collections\ArrayCollection].

Comment (h-sn) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: It returns empty collection somehow

Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: Please apply my previous comments, should help.

Comment (h-sn) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: This is the case. It returns error:

     function its_taxons_should_be_returned_by_taxonomy(TaxonInterface $taxon, TaxonInterface $rootTaxon) {

        $rootTaxon->addChild($taxon);
        $rootTaxon->setName('Brand');
        $rootTaxon->getName()->willReturn('Brand');
        $rootTaxon->getId()->willReturn(1);
        $taxon->getId()->willReturn(2);
        $taxon->getRoot()->willReturn($rootTaxon);
        $taxons = new ArrayCollection([$taxon->getWrappedObject()]);
        $this->setTaxons($taxons);
        $taxon->getRoot()->shouldBeCalled();
        $rootTaxon->getName()->shouldBeCalled();

        $returnedTaxons = $this->getTaxons('Brand');
        $returnedTaxons->shouldHaveType('Doctrine\Common\Collections\Collection');
        $returnedTaxons->shouldReturn(new ArrayCollection([$taxon]));
    }

**Comment by h-sn** [comment]
The last commit is the shortest version, that tests correctness of invoked methods without errors, I was able to do quickly.

**Comment by michalmarcinkowski** [comment]
The #5215 was merged. As I mentioned [here](https://github.com/Sylius/Sylius/pull/5215#issuecomment-225034490) @h-sn or @coudenysj are you willing to add spec to prevent from regression?

**Comment by h-sn** [comment]
I've already added the spec for this. What's wrong with it?
[last spec commit](https://github.com/Sylius/Sylius/pull/5170/commits/988393692eb317bb1dc9747eb24a20209abe3f50)

File: src/Sylius/Component/Core/spec/Model/ProductSpec.php [hunk 9]
```diff
@@ -67,6 +68,19 @@ function its_taxons_are_mutable(Collection $taxons)
...
+        TaxonInterface $taxon,
+        TaxonInterface $rootTaxon
+    ) {
+        $rootTaxon->getName()->willReturn('Brand');
+        $taxon->getRoot()->willReturn($rootTaxon);
+        $this->setTaxons(new ArrayCollection([$taxon->getWrappedObject()]));
+
+        $returnedTaxons = $this->getTaxons('brand');
+        $returnedTaxons->shouldHaveType('Doctrine\Common\Collections\Collection');
+        $returnedTaxons->count()->shouldReturn(1);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/ProductSpec.php]: `$this->getTaxons('brand')->shouldReturn(new ArrayCollection([$taxon->getWrappedObject()]));`

**Comment by michalmarcinkowski** [comment]
Please rebase and apply my last comment.

**Comment by pamil** [comment]
Thanks for your contribution! 🎉  Unfortunately it seems like it was somehow fixed in some other PR through the year.

**Pull request closed without merging**